### PR TITLE
Async functions

### PR DESCRIPTION
**Pull request by borodean**
Implements asynchronous functions by allowing to return Promises from them:

``` js
var options = {
  functions: {
    'bar': function () {
      return Promise.resolve('baz');
    }
  }
};
```

However, this breaks nested functions handling. I'll try to look into it soon.

### CODE REVIEW

**Comment by borodean** [comment]
Fixed nested functions. Now the code requires some refactoring to be tidier.

**Comment by borodean** [comment]
Hm, just noticed that the way I've updated the tests copies what @TrySound did on PR https://github.com/andyjansson/postcss-functions/pull/9.

I'll borrow his fix for node 0.10.

**Comment by borodean** [comment]
@andyjansson what do you think about this feature?

Also, I need some advice on further refactoring.

**Comment by borodean** [comment]
(Don't mind it seriously, I'm just thinking out loud on refactoring possibilities.)

For `postcss-assets` I've written a Gonzales-based function mapper: [postcss-assets/lib/map-functions.js](https://github.com/borodean/postcss-assets/blob/develop/lib/map-functions.js)

There I utilized a simpler approach to parse arguments before passing them to functions: it was just `comma` method from `postcss/lib/list` (line 47), since I was too lazy to process a tree of nodes for getting them. However, the whole thing is still pretty advanced: [postcss-assets/test/map-functions.js](https://github.com/borodean/postcss-assets/blob/develop/test/map-functions.js).

Maybe this argument parsing technique could be used for `postcss-functions` as well.

**Comment by TrySound** [comment]
@borodean He likes atomic PRs. I'd like to implement firstly async tests. async loading and then async functions. Let's wait my two PRs and then you can rebase your.

> Maybe this argument parsing technique could be used for postcss-functions as well.

Reparsing I think is more long process then just squashing parsed arguments.

**Comment by borodean** [comment]
> Let's wait my two PRs and then you can rebase your.

Sure, I'll rebase when yours are done.

What's an async loading?

**Comment by TrySound** [comment]
@borodean I meant asynchronous file detecting with globby

**Comment by andyjansson** [comment]
@borodean I've merged @TrySound's async tests and reviewed your changes. Looks good. Awaiting a rebase.

**Comment by borodean** [comment]
@andyjansson rebased. I've also squashed everything into one commit.

File: index.js [hunk 1]
```diff
@@ -4,32 +4,55 @@ var postcss = require('postcss'),
...
+				return prev + result;
+			});
 	}, '');
 	if (last) {
 		args.push(last);
 	}
 
-	return args;
+	return Promise.all(args);
+}
```
Comment (TrySound) [inline index.js:27]: Why you changed thid function? Everything here was fine.

Comment (borodean) [inline index.js:27]: @TrySound what do you mean? It has to wait until all the calls to `transformFunction` are resolved before returning the result.

**Comment by borodean** [comment]
@andyjansson really cool to have it merged! I'm going to reuse this instead of my own code in [postcss-assets](https://github.com/assetsjs/postcss-assets) plugin.

File: index.js (hunk 1 above)
Comment (TrySound) [inline index.js:27]: I understand. Sorry.

Comment (borodean) [inline index.js:27]: Nothing to be sorry about :-) Cheers!

**Comment by borodean** [comment]
@andyjansson when do you expect to have a new release published?

**Comment by andyjansson** [comment]
@borodean I can publish a new version after work.

**Comment by andyjansson** [comment]
@borodean v2.1.0 has been released.

**Comment by borodean** [comment]
@andyjansson fantastic, thanks for your effort!

**Pull request merged**